components/core/src/clp/streaming_archive/single_file_archive/writer.cpp

```diff
@@ -0,0 +1,297 @@
+ * Skips metadata file since already written in `write_archive_metadata`.
+ *
+ * @param archive_writer
+ * @param multi_file_archive_path
+ * @param segment_ids
+ * @throws Propagates `update_offset`'s exceptions.
+ */
+auto write_archive_files(
+        FileWriter& archive_writer,
+        std::filesystem::path const& multi_file_archive_path,
+        std::vector<std::string> const& segment_ids
+) -> void;
+
+auto get_segment_ids(segment_id_t next_segment_id) -> std::vector<std::string> {
+    std::vector<std::string> segment_ids;
+
+    if (next_segment_id == 0) {
+        return segment_ids;
+    }
+
+    for (size_t i = 0; i < next_segment_id; ++i) {
+        segment_ids.emplace_back(std::to_string(i));
+    }
+
+    return segment_ids;
+}
+
+auto get_file_size_and_update_offset(std::filesystem::path const& file_path, uint64_t& offset)
+        -> void {
+    try {
+        auto size = std::filesystem::file_size(file_path);
+        offset += size;
+    } catch (std::filesystem::filesystem_error const& e) {
+        throw OperationFailed(
+                ErrorCode_Failure,
+                __FILENAME__,
+                __LINE__,
+                fmt::format("Failed to get file size: {}", e.what())
+        );
+    }
+}
+
+auto get_file_infos(
+        std::filesystem::path const& multi_file_archive_path,
+        std::vector<std::string> const& segment_ids
+) -> std::vector<FileInfo> {
+    std::vector<FileInfo> files;
+    uint64_t offset = 0;
+
+    for (auto const& static_archive_file_name : cStaticArchiveFileNames) {
+        files.emplace_back(FileInfo{std::string(static_archive_file_name), offset});
+        get_file_size_and_update_offset(multi_file_archive_path / static_archive_file_name, offset);
+    }
+
+    std::filesystem::path segment_dir_path = multi_file_archive_path / cSegmentsDirname;
+    for (auto const& segment_id : segment_ids) {
+        files.emplace_back(FileInfo{segment_id, offset});
+        get_file_size_and_update_offset(segment_dir_path / segment_id, offset);
+    }
+
+    // Add sentinel indicating total size of all files.
+    files.emplace_back(FileInfo{"", offset});
+
+    // Decompression of large single-file archives will consume excessive memory since
+    // single-file archives are not split.
+    if (offset > cFileSizeWarningThreshold) {
+        SPDLOG_WARN(
+                "Single file archive size exceeded {}. "
+                "The single-file archive format is not intended for large archives, "
+                " consider using multi-file archive format instead.",
+                cFileSizeWarningThreshold
+        );
+    }
+
+    return files;
+}
+
+auto pack_single_file_archive_metadata(
+        ArchiveMetadata const& multi_file_archive_metadata,
+        std::vector<FileInfo> const& file_infos,
+        size_t num_segments
+) -> std::stringstream {
+    MultiFileArchiveMetadata archive_metadata{
```

Comment:
If we make SingleFileArchiveMetadata a class instead of a struct, is there a chance to let it directly use "ArchiveMetadata"?

I kind of feel the conversion between ArchiveMetadata to MultiFileArchiveMetadata looks redundant. 
Even if we really want to use such a intermediate struct, I feel it should be defined under ArchiveMetadata.hpp, and we should implement a member function under ArchiveMetadata class thatr returns a MultiFileArchiveMetadata/